homeassistant/components/homekit_controller/humidifier.py

```diff
@@ -0,0 +1,224 @@
+"""Support for HomeKit Controller humidifier."""
+from typing import Any, Dict, List, Optional
+
+from aiohomekit.model.characteristics import CharacteristicsTypes
+
+from homeassistant.components.humidifier import HumidifierEntity
+from homeassistant.components.humidifier.const import (
+    ATTR_AVAILABLE_MODES,
+    ATTR_HUMIDITY,
+    ATTR_MAX_HUMIDITY,
+    ATTR_MIN_HUMIDITY,
+    ATTR_MODE,
+    DEVICE_CLASS_DEHUMIDIFIER,
+    DEVICE_CLASS_HUMIDIFIER,
+    SUPPORT_MODES,
+)
+from homeassistant.core import callback
+
+from . import KNOWN_DEVICES, HomeKitEntity
+
+SUPPORT_FLAGS = 0
+
+ATTR_DEHUMIDIFIER_THRESHOLD = "dehumidifier_threshold"
+ATTR_HUMIDIFIER_THRESHOLD = "humidifier_threshold"
+ATTR_CURRENT_HUMIDITY = "current_humidity"
+
+HK_MODE_TO_HA = {
+    0: "off",
+    1: "auto",
+    2: "humidifying",
+    3: "dehumidifying",
+}
+
+HA_MODE_TO_HK = {
+    "auto": 0,
+    "humidifying": 1,
+    "dehumidifying": 2,
+}
+
+
+class HomeKitHumidifierDehumidifier(HomeKitEntity, HumidifierEntity):
+    """Representation of a HomeKit Controller Humidifier."""
+
+    def get_characteristic_types(self):
+        """Define the homekit characteristics the entity cares about."""
+        return [
+            CharacteristicsTypes.ACTIVE,
+            CharacteristicsTypes.RELATIVE_HUMIDITY_CURRENT,
+            CharacteristicsTypes.CURRENT_HUMIDIFIER_DEHUMIDIFIER_STATE,
+            CharacteristicsTypes.TARGET_HUMIDIFIER_DEHUMIDIFIER_STATE,
+            CharacteristicsTypes.RELATIVE_HUMIDITY_HUMIDIFIER_THRESHOLD,
+            CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD,
+        ]
+
+    def _char_threshold_for_current_mode(self):
+        if self.device_class == DEVICE_CLASS_DEHUMIDIFIER:
+            return CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD
+
+        if self.device_class == DEVICE_CLASS_HUMIDIFIER:
+            return CharacteristicsTypes.RELATIVE_HUMIDITY_HUMIDIFIER_THRESHOLD
+
+        return (
+            CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD
+            if self.mode == "dehumidifying"
+            else CharacteristicsTypes.RELATIVE_HUMIDITY_HUMIDIFIER_THRESHOLD
+        )
+
+    @property
+    def device_class(self) -> str:
+        """Return the device class of the device."""
+        dehumidifier_threshold = self.service.value(
+            CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD
+        )
+        humidifier_threshold = self.service.value(
+            CharacteristicsTypes.RELATIVE_HUMIDITY_HUMIDIFIER_THRESHOLD
+        )
+
+        if dehumidifier_threshold is not None and humidifier_threshold is not None:
+            return None
+
+        if dehumidifier_threshold is not None:
+            return DEVICE_CLASS_DEHUMIDIFIER
+
+        if humidifier_threshold is not None:
+            return DEVICE_CLASS_HUMIDIFIER
+
+        return None
```

Comment:
Why testing both cases ? Can it happen to return `None` ?

Maybe this is possible:
```suggestion
        return (
            DEVICE_CLASS_DEHUMIDIFIER
            if dehumidifier_threshold is not None
            else DEVICE_CLASS_HUMIDIFIER
        )
```

Also maybe store this in a variable so we don't have to compute the `dehumidifier_threshold` every time.

Reply:
The HomeKit Accessory Protocol defines both threshold characteristics as optional, so it is possible in theory.